src/zenml/config/source.py

```diff
@@ -162,6 +164,197 @@ def model_dump_json(self, **kwargs: Any) -> str:
+    serialization time
+    - Lazy loading: Only loads sources when explicitly requested
+
+    Examples:
+    * the source `SourceOrObject(module="zenml.config.source", attribute="SourceOrObject")`
+    references the class that this docstring is describing.
+    * the object `SourceOrObject.from_object(object=my_object)` creates a source
+    or object from the object `my_object`.
+
+    Attributes:
+        _object: Loaded object object (if initialized from object or loaded).
+        _is_loaded: Whether the callable has been loaded.
+    """
+
+    _object: Optional[ObjectType] = None
+    _is_loaded: bool = False
+
+    @classmethod
+    def from_source(cls, source: Source) -> "SourceOrObject":
+        """Creates a source or object from a source instance.
+
+        Args:
+            source: The source instance.
+
+        Returns:
+            The source or object instance.
+        """
+        return cls(**source.model_dump())
+
+    @classmethod
+    def from_object(cls, object: ObjectType) -> "SourceOrObject":
+        """Creates a source or object from an object instance.
+
+        Args:
+            object: The object instance.
+
+        Returns:
+            The source or object instance.
+        """
+        # We don't resolve the object right away, we only do it at serialization
+        # time. This allows us to store temporary objects in this class too.
+        result = cls(
+            module="",
+            type=SourceType.UNKNOWN,
+        )
+        result._object = object
+        result._is_loaded = True
+        return result
+
+    def load(self) -> ObjectType:
+        """Load and return the object.
+
+        Returns:
+            The loaded object.
+
+        Raises:
+            RuntimeError: If loading fails.
+        """
+        from zenml.utils import source_utils
+
+        if not self._is_loaded:
+            try:
+                self._object = source_utils.load(self)
+                self._is_loaded = True
+            except Exception as e:
+                raise RuntimeError(
+                    f"Failed to load object from {self.import_path}: {e}"
+                ) from e
+
+        assert self._object is not None
+        return self._object
+
+    def resolve(self) -> None:
+        """Resolve the stored object to a source."""
+        from zenml.utils import source_utils
+
+        if not self._is_loaded:
+            return
+
+        if self.module:
+            # Already resolved
+            return
+
+        source = source_utils.resolve(self._object)
+        self.module = source.module
+        self.attribute = source.attribute
+        self.type = source.type
+
+    @property
+    def is_loaded(self) -> bool:
+        """Whether the object has been loaded.
+
+        Returns:
+            True if the object has been loaded, False otherwise.
+        """
+        return self._is_loaded
+
+    @classmethod
+    def convert_source_or_object(cls, source: Any) -> Any:
```

Comment:
Doesn't this return an item of the class, not `Any`? If so, `-> Self` I think would be better here